### ISSUE TITLE
Provide per-node query latency percentiles, or the ability for a plugin to collect them.

### ISSUE THREAD

**Issue by ebernhardson** [elastic/elasticsearch#15412]
From the existing stats i can derive per node average query latency, but the average isn't quite as interesting. Recording the 'took' time from queries I find that our percentiles come out to, roughly: 50th: 25ms, 75th: 70ms, 95th: 180ms, 99th: 8000ms. We run a non-homogeneous cluster so i'm really interested what the impact of some of the less performant hardware is on these percentiles. It would be quite interesting to learn that 99th percentile for one set of hardware is 3s, while the older hardware is up near 8s (completely guessing).

I've looked over the existing code base and it does not look like a plugin can currently collect these things, I would probably need a way to access the timing information that currently flows through org.elasticsearch.index.search.stats.ShardSearchStats.

Collecting percentiles within ES proper, providing a configuration mechanism for a plugin to replace the ShardSearchStats, or some kind of listener on ShardSearchStats would all be acceptable solutions to the problem.

**Comment by clintongormley** [comment]
@polyfractal any thoughts?

**Comment by polyfractal** [comment]
I don't know enough about the plugin architecture to comment on how easy/hard it'd be to expose this stuff for pluggability.  But I do like the idea of natively recording percentiles.

Now that we have an implementation of HDRHistogram, that'd probably be the sketch to use here.  Relatively small memory footprint, large dynamic range.  Some open questions/problems though:

- Like all other ES stats, the percentiles would be cumulative and probably lose value over time as they saturate (e.g. thousands of fast queries will, over time, move the 90th percentile down and mask real problems)
- Memory usage: iirc a 2 sig-fig HDRHistogram uses around 30kb of memory in the worst case.  So 30kb * a dozen stats is only ~360kb.  But we'd want to put a maximum dynamic range so it doesn't grow unbounded (but I don't think that would be a problem).  3 sig-figs is a lot heavier, ~200kb per sketch
- If we wanted smaller memory footprint, we could use much less accurate Frugal Streams (<1kb)
- HDRHistogram is not thread-safe, and I think multiple threads can update these stats?  Might introduce some non-negligible contention locking the histograms, or have to maintain a histogram-per-thread and increase memory usage.
- It would add a small constant time latency to requests, since any percentile recording/sketching will necessarily be slower than just bumping a cumulative time counter.  May be negligible though?
- Practically speaking, adding percentiles will make the Stats output a lot more verbose

All said, it *would* be very useful to see distributions in the stats.  All stats that expose a cumulative `time_in_millis` would benefit.

/cc @colings86 who added HDRHistogram to the aggs...any gotchas that might trip this up?

**Comment by ebernhardson** [comment]
* I'm uncertain that cumulative percentiles will be as useful as something that provides recency. Most of the servers in my cluster have been up since mid September. Any change to query patterns made now will not show up in percentiles that take into account several months and several hundred million queries.
* The search stats at least are updated by multiple threads.
* If lock contention is an issue sampling of the query stats for use in the histogram could be viable. Getting the sampling level right could be difficult though. I might err on the side of histogram-per-thread, but i'm not sure how many threads are involved so I can't estimate the impact of that.

I understand that providing recency is a much more difficult problem with stream data, but in my opinion it is paramount to these percentiles being useful.

**Comment by polyfractal** [comment]
Yep, I agree that recency is important.  The main problem is that compact/fast streaming sketches tend to be unable to deal with recency, since the relevant time information is lost as soon as the data is added to the sketch.

We could use something like windowing or resevoir sampling, which will naturally drift with the data over time.  Downside is that they require saving some amount of the data points and are therefore a fair amount heavier memory footprint.

Some sketches naturally age data, like frugal streams.  But they aren't super accurate (in part because of their ability to drift over time).

Or we could implement something like A<sup>2</sup> Bloom Filters, where you have two (or more) active copies of the sketch and rotate between them.  I haven't seen any literature for this, but I think you could actually rig something up with a triplet-ring buffer:

In pseudocode:
```rust
histos[] = new HdrHisto[3];
current = 0;

fn addData(data) {
  // Add data point to current and current + 1
  histos[current % 2].add(data);
  histos[(current + 1) % 2].add(data);

  // After a predfined interval, rotate the histos
  if (elapsed_time > interval) {
    current += 1;
    histos[(current - 1) % 2].clear();
  }
}

// only return percentile estimates from the current sketch
fn getPercentiles() {
  return histos[current].percentiles();
}
```

Basically, by rotating the sketch, you guarantee that `current` holds `2 * interval` worth of data.  The triplet is preferred over the A<sup>2</sup> setup because it avoids a "cold" sketch when it rolls over.

Alternatively, the truly simplest way to solve this is an ability to reset all the stats, so you could just manually reset at the start of testing.

**Comment by jasontedor** [comment]
> Alternatively, the truly simplest way to solve this is an ability to reset all the stats, so you could just manually reset at the start of testing.

We've already decided in #9693 that this will not be added.

**Comment by jasontedor** [comment]
Closed by #17398
